### PR TITLE
Clarify wording and naming of credentials in CSI

### LINKED ISSUE #198: Credentials difficult for COs to expose
PR https://github.com/container-storage-interface/spec/pull/168 attempted to fix issues with credentials in the CSI spec by introducing a unique credential per call: `controller_create_credentials`, `controller_delete_credentials`, `controller_publish_credentials`, `controller_unpublish_credentials`, `node_stage_credentials`, `node_unstage_credentials`, `node_publish_credentials`, and `node_unpublish_credentials`. Furthermore it requires that `COs SHALL permit passing through the required credentials`.

A few issues:
1. ~This implies the credentials for each of these calls could be different (even if the calls are for the same volume), and that COs must allow users to set them.~ Update: on rethinking this should be possible to allow plugins to control where secrets are exposed.
2. This implies the credentials for request and response for the same call could be different, and that COs must allow users to set them.
3. Exposing this many credentials to end users is confusing (it requires them to understand the inner workings of CO to SP communications), and difficult for COs to do (where in the Kubernetes API would all of these make sense).
2. nit: The word `credentials` implies identity, but it could be used to pass, for example, a decryption key (which is not really a credential).

~Proposal:~
* ~Simplify by specifying a secret per use case instead of per operation (reducing 8 possibilities to 3):~
    1. ~`create_delete_volume_secrets`~
        * ~Grants ability to create and delete volume requests.~
        * ~Passed to `CreateVolume()` and `DeleteVolume()`.~
    2. ~`volume_secrets`~
        * ~Secrets required to access a volume, independent of any consumer.~
        * ~Example: an encryption key for volume.~
        * ~Passed to `ControllerPublishVolume()`, `ControllerUnpublishVolume()`, `NodeStageVolume()`, `NodeUnstageVolume()`, `NodePublishVolume()`, `NodeUnpublishVolume()`.~
    3. ~`workload_secrets`~
        * ~Secrets required to auth a specific consumer of a volume.~
        * ~Example: A single NFS share that permits multiple users access via different credentials.~
        * ~Passed to `NodePublishVolume()`, `NodeUnpublishVolume()`~

~As we add additional functionality to the spec (e.g. snapshots) we can add more use case specific secrets.~

Considering we want to minimize the breaking changes between 0.2 and 0.3, we should consider this for 0.2.

Update: see updated "Proposal 2" below.

### PR DESCRIPTION
**Pull request by saad-ali**
Fixes #198
Alternate to https://github.com/container-storage-interface/spec/pull/200

This PR:
1. Loosens wording to allow CO to expose secrets as they choose (1 set of secret, 4 sets of secrets, 8 sets of secrets, etc. and specify there own mapping).
2. Rename credentials to secrets.
3. Remove credentials from node unpublish and nodeunstage.

### CODE REVIEW

**Comment by saad-ali** [comment]
Thanks @jieyu.
@jdef @jdef @cpuguy83 could you PTAL?

File: csi.proto [hunk 1]
```diff
@@ -196,18 +196,24 @@ message CreateVolumeRequest {
...
+  // name of the secret (e.g. "username" or "password"), and the value
+  // contains the secret data (e.g. "bob" or "abc123").
+  // Each key MUST consist of alphanumeric characters, '-', '_' or '.'.
+  // Each value MUST contain a valid string. An SP MAY choose to accept
+  // binary (non-string) data by using a binary-to-text encoding scheme,
+  // like base64.
+  // An SP SHALL advertise the requirements for required secret keys and
+  // values in documentation.
+  // CO SHALL permit passing through the required secrets. CO MAY expose
+  // secrets as a one or more map to its end users, therefore the keys
```
Comment (cpuguy83) [inline csi.proto]: This wording is a bit funky to me.
What about something like this (assuming I'm understanding what's written):

> A CO MAY send the same secrets to all RPCs, therefore the keys for all unique secrets that an SP expects must be unique across.

**Comment by saad-ali** [comment]
@cpuguy83 Reworded. PTAL

**Comment by saad-ali** [comment]
Whoops fixed!

**Pull request merged**